aten/src/ATen/native/cuda/SoftMax.cu

```diff
@@ -1005,8 +1008,71 @@ Tensor masked_softmax_cuda(const Tensor& input, const Tensor& mask) {
           mask.data_ptr<bool>()
         );
       });
-    }
-    return output;
+  }
+  return output;
+}
+
+Tensor masked_softmax_backward_cuda(
+    const Tensor& grad_,
+    const Tensor& output_,
+    int64_t dim,
+    const Tensor& mask_) {
+  Tensor grad_input = at::empty_like(grad_, grad_.options());
+  if (grad_.numel() == 0) {
+    return grad_input;
+  }
+
+  auto grad = grad_.contiguous();
+  auto output = output_.contiguous();
+  auto mask = mask_.contiguous();
+
+  grad = grad.dim() == 0 ? grad.view(1) : grad;
+  mask = mask.dim() == 0 ? mask.view(1) : mask;
+  output = output.dim() == 0 ? output.view(1) : output;
+
+  TORCH_CHECK(dim >=0 && dim < grad.dim(), "dim must be non-negative and less than input dimensions");
+  TORCH_CHECK(grad.sizes() == mask.sizes(), "Mask shape should match grad shape");
+  TORCH_CHECK(mask.scalar_type() == ScalarType::Bool, "Mask should be a boolean tensor");
+
+  int softmax_elements = output.size(dim);
+  int64_t batch_count = grad.numel() / softmax_elements;
+
+  if (softmax_elements > 1024 || softmax_elements * grad.element_size() > 4096 || dim < grad.dim()-1) {
+    AT_DISPATCH_FLOATING_TYPES_AND2(
+      ScalarType::Half,
+      ScalarType::BFloat16,
+      grad_input.scalar_type(),
+      "masked_softmax",
+      [&] {
+        Tensor mask_not = mask.logical_not();
+        grad_input = at::_softmax_backward_data(
+          grad.masked_fill(mask_not, 0),
+          output,
+          dim,
+          grad_input.scalar_type()
+        );
+      });
+  } else {
+    grad = grad * output;
+    AT_DISPATCH_FLOATING_TYPES_AND2(
+      ScalarType::Half,
+      ScalarType::BFloat16,
+      grad_input.scalar_type(),
+      "masked_softmax_backward",
+      [&] {
+        using accscalar_t = acc_type<scalar_t, true>;
+        dispatch_softmax_backward<scalar_t, scalar_t, accscalar_t, false, true /* masked_softmax */>(
+          grad_input.data_ptr<scalar_t>(),  // gI_ptr
+          grad.data_ptr<scalar_t>(),  // grad_ptr
+          output.data_ptr<scalar_t>(),  // output_ptr
+          softmax_elements,  // softmax_elements
+          softmax_elements,   // softmax_elements_stride
+          batch_count,  // batch_count
+          mask.data_ptr<bool>()  /* not masked */
+        );
+      });
+  }
+  return at::nan_to_num(grad_input);
```

Comment:
Not entirely confident this is the right approach - I want to double check the logic here. Consider the following example from `test_masked_softmax_forward_with_nans`:

```
input
tensor([[[[-0.0883,  0.3420,  0.4112,  1.0051],
          [-0.1117, -0.5988, -0.0982, -0.3511],
          [ 0.7209, -0.2169, -1.0427,  0.2448],
          [-0.9887, -0.5196,  0.6585,  0.6406]]]], device='cuda:0',
       requires_grad=True)
mask
tensor([[[[False,  True, False,  True],
          [False,  True, False,  True],
          [False,  True, False,  True],
          [False,  True, False,  True]]]], device='cuda:0')
result
tensor([[[[nan, 1., nan, 1.],
          [nan, 1., nan, 1.],
          [nan, 1., nan, 1.],
          [nan, 1., nan, 1.]]]], device='cuda:0',
       grad_fn=<MaskedSoftmaxBackward0>)
expected
tensor([[[[nan, 1., nan, 1.],
          [nan, 1., nan, 1.],
          [nan, 1., nan, 1.],
          [nan, 1., nan, 1.]]]], device='cuda:0', grad_fn=<SoftmaxBackward0>)
```

This is all expected, but the `at::nan_to_num` is necessary in order to match the gradient:
```
input_ref.grad
tensor([[[[0., 0., 0., 0.],
          [0., 0., 0., 0.],
          [0., 0., 0., 0.],
          [0., 0., 0., 0.]]]], device='cuda:0')
```
I checked the logic and even compared all the values being set in the backwards gradient from `Softmax.cpp` (where I also applied the same `nan_to_num` function, and they were identical. I'm wondering if the structured kernel takes care of this somehow by automatically converting nan's in the backwards gradient to 0?

Reply:
I don't think we ever want to override nans tbh.
In the test, if you generate nan value, well that's what you get in the end.

In the masked semantic, if these values are masked anyway, we don't care what they are and should remain whatever (nan in this case?).

I would also stress again that finite difference and autograd in general only works for differentiable ops. If you're doing non-differentiable stuff, you should not try to match these together.